### ISSUE TITLE
Add support for TestSwarm maybe?

### ISSUE THREAD

**Issue by jayarjo** [ryanseddon/bunyip#5]
Have you thought of adding testswarm support to bunyip, in addition to yeti, which as we found here seems to have some drawbacks in our usage case. Specifically the fact that it doesn't seem to support php on server-side (I wonder if it can serve nodejs files at least) and we sometimes need to interchange messages with the server and test dynamic response, so absense of php support is kinda deal-breaking.

I know you said testswarm is more after-commit oriented and remember Reid Burke saying the same in his presentation of yeti, but then it's not. They have nice API and one is not constrained to when he wants to test anything, before commit, after commit or just on request. There's nodejs module for <a href="https://github.com/jzaefferer/node-testswarm">testswarm</a> and an <a href="https://github.com/clarkbox/testswarm-browserstack">effort to integrate testswarm and browserstack</a>.

**Comment by ryanseddon** [comment]
Yeah that sounds like a great option. If you want to tackle it that'd be awesome otherwise I'll add it to the todo list.

**Comment by jayarjo** [comment]
I'm trying to wrap up something at the moment. Let's see what will come out of it.

**Comment by jayarjo** [comment]
Hey Ryan, it took a bit longer and went much further then I initially anticipated. But the whole thing is more modular now, so that for example, one can easily add other custom testrunners, tunnelers and browser provider services. It is still a bit raw, but then because the changes were so dramatic, before proceeding any further, I decided to wrap all the new stuff, modifications and rewrites quickly and expose it for some feedback from you, and community maybe.

Here's the merge of everything I did so far: b8d0de4, with some insightful comments. Will be glad to hear everyone's thoughts and suggestions.

**Comment by ryanseddon** [comment]
Wow this is some incredible stuff :heart:'z 

I will take a more detailed look but what Ive seen so far is pretty solid, i'd be happy to pull this in as a pull request. You've pretty much done everything I wanted to do on my [roadmap wiki](https://github.com/ryanseddon/bunyip/wiki/Roadmap---wanted-features).

**Comment by jayarjo** [comment]
Great that we were thinking along the same lines! This thing still requires some effort though (dependencies, tests, grunt), but it's quite straightforward now. One big obstacle though, official TestSwarm still doesn't provide detailed test results in json format (like what failed and where), this one is deal-breaking I guess, but as far as I understood TestSwarm developer is also interested in adding that bit, and then I made a pull request here: https://github.com/jquery/testswarm/issues/183. Waiting for a response.
